contributors/design-proposals/secure_watch.md

```diff
@@ -0,0 +1,288 @@
+BulkSelector above).
+
+ With the above assumption, what we can guarantee is:
+1. within a single resource type, watch events are delivered in increasing
+order of resource version
+1. within a single resource type, there is at most one watch event delivered
+for any resource version
+
+ On top of the following assumptions, we can simply reuse already existing
+frameworks for bulk watches of objects of only single resource type.
+To support multi-resource-type watches, we can build another framework on top
+of what we already have as following:
+- we will have a single reflector/informer/... per resource type
+- we will add an aggregator/dispatcher in front of them, that will be
+responsible for aggregating requests from underlying frameworks into a single
+one and then dispatching incoming watch events to the correct one.
+ This will obviously require some changes to the existing frameworks, but those
+shouldn't be that big.
+
+ One more thing to mention is detecting resource type of object being send via
+watch. With bulk watch, you don't really know object of what type has just been
+sent to you.
+ However, it seems that we already have all the logic in place to detect object
+type by deserializing only parts of it.
+
+TODO: We should check if it really works for protobufs. If not (and there won't
+be any relatively easy way to make it work), I would suggest introducing a
+dedicated type as following:
+```
+type BulkWatchEvent struct {
+	Type watch.EventType
+
+	// FIXME: This should be meta.k8s.io/GroupVersionKind instead (or something
+	// similar to it).
+	ResourceType string
+
+	Object runtime.Object
+}
+```
+But hopefully, this shouldn't be needed.
+
+
+### Implementation details
+
+ The implementation of bulk get and bulk watch will work as an aggregator.
+Whenever a request is coming, it will:
+- check what resource types are requested in this request
+- for every resource type, combine only parts of the filter that are about
+this particular resource type and send exactly one request down the stack
+- gather responses for all those resource types and combine them into single
+response to the user
+
+ The only non-trivial operation above is sending the request for a single
+resource type down the stack. In order to implement it, we will need to
+slighly modify the interface of "Registry" in apiserver. The modification
+will have to allow passing both what we are passing now and BulkSelector
+(in some format) (this may e.g. changing signature to accept BulkSelector
+and translating ListOptions to BulkSelector in the current code).
+
+ With those changes in place, there is a question of how to call this
+code. There are two main options:
+1. Make each registered resource type, register also in BulkAggregator
+(or whatever we call it) and call those methods directly
+1. Expose also per-resource bulk operation in the main API, e.g.:
+```
+POST /api/v1/namespace/default/pods/bulkwatch <body defines filtering>
```

Comment:
This collides with pods called `bulkwatch` so we can't do it.  Other options might be `_bulkwatch` but there is no guarantee that a particular resource doesn't support underscore in names.  In general, I don't want to confuse request patterns (/foo/bar/?/baz should never collide with another route), so I believe option 2 is not worth pursuing.